File: android/app/src/main/java/sh/mo/SlashPolicy.kt
package sh.mo

/**
 * T10b composer slash policy (PI_TASK_T10B "Behaviour", closed decisions),
 * pure Kotlin — no Android imports, JVM-tested. The composable renders what
 * it is handed; the view model and repository call these answers.
 */
object SlashPolicy {

    /** Commands the phone answers locally, never `core.send` / `run_slash`. */
    private val LOCAL_COMMANDS = setOf("/clear", "/sessions", "/quit")

    /**
     * Decision 1: complete on a leading `/` with NO space (the argument stage
     * is out of scope). Empty draft hides.
     */
    fun shouldComplete(draft: String): Boolean =
        draft.startsWith("/") && !draft.contains(' ')

    /**
     * Decision 2: a picked row rewrites the draft from `replaceFrom` (1 when
     * `-1`/absent) and adds a trailing space, then the popup hides.
     * `replaceFrom` is the 1-based character position the replacement starts
     * at (Desktop's use-slash-completions.ts: `text.slice(0, replaceFrom)` is
     * the kept prefix, so the kept length is `replaceFrom - 1`).
     */
    fun insertCompletion(draft: String, itemText: String, replaceFrom: Long): String {
        val from = if (replaceFrom < 1) 1L else replaceFrom
        val keepEnd = (from - 1).toInt().coerceIn(0, draft.length)
        return draft.take(keepEnd) + itemText + " "
    }

    /**
     * Decision 3: `/clear` `/sessions` `/quit` are local. The match is the
     * whole token — `/clears` or `/clear now` are NOT the local command.
     */
    fun isLocalCommand(draft: String): Boolean = draft.trim() in LOCAL_COMMANDS

    /** Decision 3: a leading `/` routes the submit through the slash ladder. */
    fun isSlashSubmit(draft: String): Boolean = draft.startsWith("/")
}

/** One popup row: the DTO fields the phone renders (T10b decision 2). */
data class SlashCompletionRow(
    val text: String,
    val display: String,
    val kind: String,
    val meta: String,
)

File: ios/Sources/HermoLogic/SlashPolicy.swift
import Foundation

public enum SlashPolicy {

    private static let localCommands: Set<String> = ["/clear", "/sessions", "/quit"]

    /// The argument stage is out of scope: a space after the leading slash hides the popup.
    public static func shouldComplete(_ draft: String) -> Bool {
        draft.hasPrefix("/") && !draft.contains(" ")
    }

    /// `replaceFrom` is 1-based; a missing or negative value means replace the whole token.
    /// The position counts UTF-16 units, the unit the gateway's own `text.slice` uses.
    public static func insertCompletion(_ draft: String, _ itemText: String, _ replaceFrom: Int64) -> String {
        let from = replaceFrom < 1 ? 1 : replaceFrom
        let units = draft.utf16
        let keepEnd = min(max(Int(clamping: from) - 1, 0), units.count)
        let prefix = String(decoding: Array(units.prefix(keepEnd)), as: UTF16.self)
        return prefix + itemText + " "
    }

    /// The match is the whole trimmed token, so "/clears" or "/clear now" is not a local command.
    public static func isLocalCommand(_ draft: String) -> Bool {
        localCommands.contains(draft.trimmingCharacters(in: .whitespacesAndNewlines))
    }

    public static func isSlashSubmit(_ draft: String) -> Bool {
        draft.hasPrefix("/")
    }
}

public struct SlashCompletionRow: Equatable, Sendable {
    public let text: String
    public let display: String
    public let kind: String
    public let meta: String

    public init(text: String, display: String, kind: String, meta: String) {
        self.text = text
        self.display = display
        self.kind = kind
        self.meta = meta
    }
}
